Generalizescript

mvkski opened the pull request:
Generalize pdf generation function, so it can be consumed by various projects at code-dot-org

mvkski committed:
Generalize generatePdf scripts for multiple usecases

Write a general generatePdf function wrapping puppeteer, with generatePdfCli
as one consumer of generatePdf. Add option to return raw pdf contents to
standard out.

mvkski committed:
Add unit tests for pdf generation function, cli validation

mvkski committed:
Move package main to index.js, to export only functions needed outside package

uponthesun commented on generate-pdf.js:
```diff
@@ -0,0 +1,97 @@
+      type: "boolean",
+      describe: "if present, contents of the PDF will be written to standard out"
+    })
+    .option("optionOverrides", {
```
Typing json in the command line is pretty cumbersome. Can we look into other options, perhaps having yargs put all unknown arguments into the option overrides? Then you could just type them like regular CLI args. If that seems too unstructured, maybe only selecting unknown arguments with a certain prefix, or something like that.

mvkski replied:
Definitely worth looking into! I really like the idea of having yargs read all unknown parameters as pdf printing options - my one concern with it is that the options parameter can have nested parameters, and I dont know if this idea can easily facilitate that. For example, in the default parameters

```
DEFAULT_PRINT_OPTIONS = {                                                        
  format: "Letter",                                                              
  margin: {                                                                      
    top: "0.5in",                                                                
    bottom: "0.5in",                                                             
    right: "0.5in",                                                              
    left: "0.5in"                                                                
  }                                                                              
}                                                                               
```

`top`, `bottom`, `right`, and `left` are all underneath `margin`. 

How would you feel about having the tool point to a file which contains override parameters?

uponthesun replied:
Good point, reading from a file sounds good.

uponthesun commented on generate-pdf.js:
```diff
@@ -0,0 +1,97 @@
+  return await page.pdf(pdfOptions)
+}
+
+function validatePdfCliArgs(argv) {
```
Some of the validation here seems redundant with the validation features yargs already has, like marking options as required. Can we use the built in features where possible?

mvkski replied:
I explicitly removed using yarg's demandOption feature in favor of what we have here. I had thought about it, and I have some reasons why I believe it is better as written

Advantages of what is here:
* All input errors are reported to the user, no matter what combination of input errors are supplued. Using demandOption, if an option listed in demandOptions is missing yargs will fail with an error demanding the option be supplied without calling the check method and will thus not report on other input errors.
* When reading the script, there is one and only one function/section of code to read which does all validation. I think it makes it easier for someone reading the script to not miss what validation is being done.

If the only validation needed was demanding the presence of certain options, I would 100% support using demandOption. But, seeing as we have a mix of different kind of validations that need to happen, I think its better to put them all in the same place.

uponthesun replied:
Okay, sounds good.

mvkski committed:
Change optionOverrides param to take in location of json file rather than raw json

Also add some test fixtures to test this functionality.

uponthesun commented on generate-pdf.js:
```diff
@@ -0,0 +1,113 @@
+    })
+    .option("optionOverridesFile", {
+      alias: "o",
+      describe: "path to json file contianing overrides to default pdf print parameters"
```
typo (containing)

mvkski replied:
thanks for the catch

uponthesun commented on generate-pdf.js:
```diff
@@ -73,16 +84,21 @@ async function getUrlPdfCli() {
+      browser,
+      argv.url,
+      argv.writePath,
+      argv.optionOverrides && readJsonFile(argv.optionOverrides)
```
What is `readJsonFile`? (`parseJsonFile`?)

mvkski replied:
it should be parseJsonFile. Missed this because there is no unit/integration test for the cli function itself

mvkski committed:
Add end to end tests, change directory structure

Added end to end tests for cli after an untested typo was pointed out in the
cli function.  Add directory structure to organize growing number of files

uponthesun reviewed (approved):
Looks great!
